docs/guides/contexts_expressions.md

```diff
@@ -0,0 +1,390 @@
+the global `settings`.  Because contexts are `ChainMaps`, they can all access
+the settings:
+
+```python
+In [4]: ctx['settings.appearance.theme']
+Out[4]: 'dark'
+```
+
+When we evaluate an expression, we can provide it one of these context objects:
+
+```python
+In [5]: expr = LayerListContextKeys.layers_selection_count > 0
+
+In [6]: expr.eval(ctx)
+Out[6]: False
+```
+
+### Updating Contexts
+
+You may be wondering exactly how objects such as `Viewer` and `LayerList` update
+the keys in their contexts. The aforementioned
+[`ContextNamespace`](#contextkey-objects) comes into play here again.  A
+`ContextNamespace` can be instantiated, and bound to a specific `Context`
+instance.
+
+```python
+In [6]: ctx = get_context(viewer.layers)
+
+In [7]: llck = LayerListContextKeys(ctx)
+```
+
+Attributes of an instantiated `ContextNamespace` now act as getters (and
+setters!) of their respective `ContextKey` in the associated `Context`.
+
+```python
+In [8]: llck.layers_selection_count
+Out[8]: 0
+
+In [9]: ctx['layers_selection_count']
+Out[9]: 0
+```
+
+Finally, the `update` method can be used to trigger an update of all of the
+`ContextKeys` in a given `ContextNamespace` whenever a specific event occurs,
+(the value is updated by passing the `event.source` to the `getter` function
+that was declared in the [`ContextKey` constructor](#contextkey-objects))
+
+The following example works because *all* of the `getter` methods declared
+in `LayerListContextKeys` take an instance of `layers.selection`:
+
+```python
+In [9]: viewer.layers.selection.events.changed.connect(llck.update)
+
+In [10]: viewer.add_points()
+Out[10]: <Points layer 'Points' at 0x13c62b1f0>
+
+In [11]: llck.layers_selection_count
+Out[11]: 1
+
+In [12]: ctx['layers_selection_count']
+Out[12]: 1
+```
+
+## Summary: A (rough) full picture
+
+1. napari creates special context "names" using `ContextKey` and `ContextNamespace`
+
+    ```python
+    class LayerListContextKeys(ContextNamespace):
+        active_layer_type = ContextKey(
+            None,
+            "Lowercase name of active layer type, or None of none active.",
+            lambda s: s.active and s.active._type_string
+        )
+        active_layer_is_rgb = ContextKey(
+            False,
+            "True when the active layer is RGB",
+            lambda s: getattr(s.active, "rgb", False)
+        )
+    ```
+
+2. _Internally_ (in napari code), we can use those objects directly to declare
+   expressions in an IDE-friendly way.  For example, here we are declaratively
+   populating the layer-list context menu; this is a function that will split
+   the current stack into multiple layers, but it is only enabled when the
+   selected image is a (non-RGB) `Image` layer.
+
+    ```python
+    'napari:split_stack': {
+        'description': trans._('Split Stack'),
+        'action': _split_stack,
+        'enable_when': LLCK.active_layer_type == "image",
+        'show_when': ~LLCK.active_layer_is_rgb,
+    }
+    ```
+
```

Comment:
I think there might be one point which might not getting through in the full document, is that this is necessary if we want to be able to update the value of the expressions when the values are changing, and why we don't use direct Python object. Maybe in the motivation at the top?